### ISSUE TITLE
Can Saule support relationships that have Ids without an include?

### ISSUE THREAD

**Issue by jrexdepew** [joukevandermaas/saule#58]
I am currently using your library to upgrade an existing ember/asp.netweb api project from the rest adapter to the json api adapter.  We have a relationship in our code that is one to many.  When the api is called for the parent resource, the response only contains id's of the child objects and does not sideload them.  Once ember encounters only the ids of the children it then async loads them.  When setting up this relationship in saule it seems that either the ids of the children must be excluded from the payload completely or they must be sideloaded.  I was able to fix this by removing the negation of the IsResourceIncluded call on line 250 of ResourceSerializer.cs but this removes the sideloading functionality completely.

``` c#
if (IsResourceIncluded(includedData))
{
    _includedSection.Add(includedData);
}
```

From what I understand from the spec here: http://jsonapi.org/format/#fetching-relationships
 I should be able to have a response that only includes ids of the child objects in the parent objects relationship data section.

For example if I have the object Articles and it has many Tags.  In c# my Articles class may look like:

``` C#
public class Articles {
    public string Id
    public string Title
    public List<String> TagIds
}
```

Then my Articles Resource might look like:

``` C#
public class ArticlesResource() : ApiResource
{

    public ArticleResource()
    {
      Attribute("Title");
      HasMany<TagResource>("TagIds");
     }
}
```

My return from a call to the ArticleResource endpoint could look like

``` JSON
{
  "links": {
    "self": "http://example.com/articles/1"
  },
  "data": {
    "type": "articles",
    "id": "1",
    "attributes": {
      "title": "JSON API paints my bikeshed!"
    },
    "relationships": {
      "tags": {
        "links": {
          "related": "http://example.com/articles/1/tags"
        },
        "data": [
            { "type": "tags", "id": "2" },
            { "type": "tags", "id": "3" }
         ]
      }
    }
  }
}
```

If I set this up in Salue then my payload from the articles endpoint will always have an included section.  Since ember then thinks it has the full objects from tags it will not fetch the tags from their tag endpoint.

I feel like I am either confused about how the relationships and payloads should be setup or that Saule does not allow this pattern from looking at the source.  If more clarity is needed on my problem let me know and I can screen share or send some more example code.

**Comment by joukevandermaas** [comment]
Currently Saule supports two scenarios. Either you provide the entire related resource as a property (`List<Tag>` in this case) or you don't provide it at all (`null` or missing property).

In the first case, Saule adds the `included` section, in the second it only serializes the links.

If I understand correctly, you want to provide just the ids, without including the corresponding data. This scenario is currently not supported, but I can see the usefulness. I will see if I can add something like this soon.

Thank you for reporting this!
________________________________

**Comment by jrexdepew** [comment]
Yes, you understand correctly.  Thank you for such a prompt response!  I believe for now I am able to change our code to match one of the scenarios you have listed.

**Comment by goo32** [comment]
According to the spec, if you're not including resources in the payload then you should only supply the related links. If the payload includes related resources then it MUST supply the 'resource identifier objects'. Since Saule defaults to include related resources when they are present (ie. defined in the resource model) then it should include them even when null. Currently Saule does not include null resource identifier objects. I believe this should be added as a separate issue.

**Comment by joukevandermaas** [comment]
See #125 and #126.

**Comment by goo32** [comment]
In reference to the original question...

> From what I understand from the spec here: http://jsonapi.org/format/#fetching-relationships
> I should be able to have a response that only includes ids of the child objects in the parent objects relationship data section.

Correct me if I'm wrong but the examples in the spec that only include id's are the relationship endpoints, not the parent endpoint. If you do specify relationship ids in the parent's response (ie. via "resource identifier objects"), they MUST reference either primary data or included data (presumably within the same response).

**Comment by jrexdepew** [comment]
I believe that the JSON API spec is not as specific as you are stating.  My experience is with ember data which is the main reference implementation of the JSON API spec.  If you look here: http://emberjs.com/api/data/classes/DS.EmbeddedRecordsMixin.html you will see that ember data expects ids by default in its serializer but you must specify if you are embedding the actual record in the payload.  This leads me to believe that both ways are supported and as such your api should be able to support them.  Without being able to serialize as ids in your payload, you are not really able to lazy load in the most optimal way and you would need to make additional relationship calls directly.

**Comment by goo32** [comment]
You are correct, the spec is not super clear about this. However, the DS.EmbeddedRecordsMixin is not used in the DS.JSONAPISerializer. In fact there's a [test](https://github.com/emberjs/data/blob/v2.7.0/tests/integration/serializers/json-api-serializer-test.js#L289) to warn if it is used. Maybe you're on an older version of ember data?

Either way, it's probably worth investigating further before implementing this in Saule.

**Comment by joukevandermaas** [comment]
I think this is resolved with #126.
